[cairo] Use the serializer generator for WebCore::Font, WebCore::FontPlatformData, etc

fujii opened the pull request:
#### 4a83022480ea57d4b3bb1075dee195edcaa75740
<pre>
[cairo] Use the serializer generator for WebCore::Font, WebCore::FontPlatformData, etc
<a href="https://bugs.webkit.org/show_bug.cgi?id=284025">https://bugs.webkit.org/show_bug.cgi?id=284025</a>

Reviewed by Don Olmstead.

Added serializer definitions for WebCore font related classes of
Windows Cairo port. Added a wrapper class CoreIPCLOGFONT for LOGFONT.
Removed the handwritten coders.

These code will be removed after the Skia migration.

* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
(WebCore::FontCustomPlatformData::serializedData const):
* Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
(WebCore::FontPlatformData::fromIPCData):
(WebCore::FontPlatformData::toIPCData const):
* Source/WebCore/platform/graphics/win/FontCustomPlatformDataWin.cpp:
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
(WebCore::FontCustomPlatformData::serializedData const):
* Source/WebCore/platform/graphics/win/FontPlatformDataWin.cpp:
(WebCore::FontPlatformData::fromIPCData):
(WebCore::FontPlatformData::toIPCData const):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformPlayStation.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;WebCore::Font&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;Font&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;WebCore::FontCustomPlatformData&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;FontCustomPlatformData&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;WebCore::FontPlatformDataAttributes&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;FontPlatformDataAttributes&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreFont.serialization.in:
* Source/WebKit/Shared/cairo/WebCoreFontCairo.serialization.in: Added.
* Source/WebKit/Shared/freetype/WebCoreArgumentCodersFreeType.cpp: Removed.
* Source/WebKit/Shared/win/CoreIPCLOGFONT.h: Added.
(WebKit::CoreIPCLOGFONT::CoreIPCLOGFONT):
(WebKit::CoreIPCLOGFONT::create):
(WebKit::CoreIPCLOGFONT::data const):
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp: Removed.
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):

Canonical link: <a href="https://commits.webkit.org/287413@main">https://commits.webkit.org/287413@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a49c8bf52608787f5f8661635364d8a9f0f3e65

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58598 "Built successfully") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32991 "") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84157 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30665 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81720 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6889 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62236 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20092 "Passed tests") 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82663 "Failed to checkout and rebase branch from PR 37414") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52290 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/32991 "") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42544 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49629 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/32991 "") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29086 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70765 "") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/32991 "") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85569 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6858 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4790 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70484 "Passed tests") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7029 "Failed to checkout and rebase branch from PR 37414") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/32991 "") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69729 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/55/builds/32991 "") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12291 "Built successfully and passed tests") | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6810 "") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6710 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10196 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8507 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

donny-dont reviewed (approved):
LGTM. Just one question for you. Would it make sense to add a `Platform/Cairo.cmake` to the setup? That way we can just delete that file completely after the migration. Its ok to do here or in a separate patch.

achristensen07 commented on Source/WebKit/Shared/cairo/WebCoreFontCairo.serialization.in:
```diff
@@ -0,0 +1,51 @@
+#if PLATFORM(WIN)
+headers: "CoreIPCLOGFONT.h"
+[Wrapper=WebKit::CoreIPCLOGFONT, CreateUsing=create, Alias=struct LOGFONTW, Nested] alias LOGFONT {
+    std::span<const LOGFONT> data();
```
Why is this a span if it has to have 1 and only 1 element?  Can't it just be a LOGFONT?

fujii replied:
We can't use LOGFONT here because this is the LOGFONT serializer.
I know this is a bit tricky. But, we will remove this code soon after Windows port finish the Skia migration.
If we have more types like this, I think ArgumentCoder should atomatically use encodeObject & decodeObject for is_trivially_copyable types.

fujii replied:
Another approach is listing members here.
~~~
  LONG lfHeight;
  LONG lfWidth;
  LONG lfEscapement;
  LONG lfOrientation;
  LONG lfWeight;
  BYTE lfItalic;
  BYTE lfUnderline;
  BYTE lfStrikeOut;
  BYTE lfCharSet;
  BYTE lfOutPrecision;
  BYTE lfClipPrecision;
  BYTE lfQuality;
  BYTE lfPitchAndFamily;
  std::span<const CHAR> lfFaceName();
~~~
However `std::span` again because the serializer generator don't understand `lfFaceName[LF_FACESIZE]` yet.

fujii replied:
Ah. We have `ArgumentCoder<std::array<T, size>>`. Will try it.

fujii replied:
It's complicated than I expected. I will try it in follow-ups.

fujii replied:
I realized `ArgumentCoder<std::span<T, Extent>` supports fixed Extent. I can use `std::span<const LOGFONT, 1>` to remove `RELEASE_ASSERT(data.size() == 1)`. I will land this at the moment.

fujii committed:
[cairo] Use the serializer generator for WebCore::Font, WebCore::FontPlatformData, etc
https://bugs.webkit.org/show_bug.cgi?id=284025

Reviewed by Don Olmstead.

Added serializer definitions for WebCore font related classes of
Windows Cairo port. Added a wrapper class CoreIPCLOGFONT for LOGFONT.
Removed the handwritten coders.

These code will be removed after the Skia migration.

* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
(WebCore::FontCustomPlatformData::serializedData const):
* Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
(WebCore::FontPlatformData::fromIPCData):
(WebCore::FontPlatformData::toIPCData const):
* Source/WebCore/platform/graphics/win/FontCustomPlatformDataWin.cpp:
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
(WebCore::FontCustomPlatformData::serializedData const):
* Source/WebCore/platform/graphics/win/FontPlatformDataWin.cpp:
(WebCore::FontPlatformData::fromIPCData):
(WebCore::FontPlatformData::toIPCData const):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformPlayStation.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<WebCore::Font>::encode): Deleted.
(IPC::ArgumentCoder<Font>::decode): Deleted.
(IPC::ArgumentCoder<WebCore::FontCustomPlatformData>::encode): Deleted.
(IPC::ArgumentCoder<FontCustomPlatformData>::decode): Deleted.
(IPC::ArgumentCoder<WebCore::FontPlatformDataAttributes>::encode): Deleted.
(IPC::ArgumentCoder<FontPlatformDataAttributes>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreFont.serialization.in:
* Source/WebKit/Shared/cairo/WebCoreFontCairo.serialization.in: Added.
* Source/WebKit/Shared/freetype/WebCoreArgumentCodersFreeType.cpp: Removed.
* Source/WebKit/Shared/win/CoreIPCLOGFONT.h: Added.
(WebKit::CoreIPCLOGFONT::CoreIPCLOGFONT):
(WebKit::CoreIPCLOGFONT::create):
(WebKit::CoreIPCLOGFONT::data const):
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp: Removed.
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):

Canonical link: https://commits.webkit.org/287413@main
